programs/lwdnsq/lwdnsq.8
'\" t
.\"     Title: IPSEC LWDNSQ
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.75.2 <http://docbook.sf.net/>
.\"      Date: 10/06/2010
.\"    Manual: [FIXME: manual]
.\"    Source: [FIXME: source]
.\"  Language: English
.\"
.TH "IPSEC LWDNSQ" "8" "10/06/2010" "[FIXME: source]" "[FIXME: manual]"
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
lwdnsq \- lookup items in DNS to help pluto (and others)
.SH "SYNOPSIS"
.HP \w'\fBipsec\fR\ 'u
\fBipsec\fR \fIlwdnsq\fR [\-\-help] [\-\-prompt] [\-\-serial] [\-\-debug] [\-\-log\ \fIlogfile\fR] [\-\-regress] [\-\-ignoreeof]
.br

.SH "DESCRIPTION"
.PP
The
\fBipsec lwdnsq\fR
is a helper program that does DNS lookups for other programs\&. It implements an asynchronous interface on stdin/stdout, with an ASCII driven command language\&.
.PP
If stdin is a tty or if the
\fB\-\-prompt\fR
option is given, then it issues a prompt to the user\&. Otherwise, it is silent, except for results\&.
.PP
The program will accept multiple queries concurrently, with each result being marked with the ID provided on the output\&. The IDs are strings\&.
.PP
If the
\fB\-\-serial\fR
option is given, then the program will not attempt to execute concurrent queries, but will serialize all input and output\&.
.SH "QUERY LANGUAGE"
.PP
There are twelve commands that the program understands\&. This is to lookup different types of records in both the forward and reverse maps\&. Every query includes a queryid, which is returned in the output, on every single line to identify the transaction\&.
.SS "HELP"
.PP
This request lists the available commands
.SS "QUIT"
.PP
This request causes lwdnsq to quit\&.
.SS "DEBUG"
.PP
This request enables debugging\&.
.SS "KEY queryid FQDN"
.PP
This request looks up the KEY resource record for the given
\fBFQDN\&.\fR\&.
.SS "KEY4 queryid A\&.B\&.C\&.D"
.PP
This request looks up the KEY resource record found in the reverse map for the IP version 4 address
\fBA\&.B\&.C\&.D\fR, i\&.e\&. it looks up D\&.C\&.B\&.A\&.in\-addr\&.arpa\&.
.SS "KEY6 queryid A:B::C:D"
.PP
This request looks up the KEY resource record found in the reverse map for the IPv6 address
\fBA:B::C:D\fR, i\&.e\&. it looks the 32\-nibble long entry in ip6\&.arpa (and ip6\&.int)\&.
.SS "TXT queryid A\&.B\&.C\&.D"
.PP
This request looks up the TXT resource record found[?]
.SS "TXT4 queryid A\&.B\&.C\&.D"
.PP
This request looks up the TXT resource record found in the reverse map for the IP version 4 address
\fBA\&.B\&.C\&.D\fR, i\&.e\&. it looks up D\&.C\&.B\&.A\&.in\-addr\&.arpa\&.
.SS "TXT6 queryid A:B::C:D"
.PP
This request looks up the TXT resource record found in the reverse map for the IPv6 address
\fBA:B::C:D\fR, i\&.e\&. it looks the 32\-nibble long entry in ip6\&.arpa (and ip6\&.int)\&.
.SS "KEY queryid FQDN"
.PP
This request looks up the IPSECKEY resource record for the given
\fBFQDN\&.\fR\&. See note about IPSECKEY processing, below\&.
.SS "IPSECKEY queryid A\&.B\&.C\&.D"
.PP
This request looks up the IPSECKEY resource record found in the reverse map[?]
.SS "IPSECKEY4 queryid A\&.B\&.C\&.D"
.PP
This request looks up the IPSECKEY resource record found in the reverse map for the IP version 4 address
\fBA\&.B\&.C\&.D\fR, i\&.e\&. it looks up D\&.C\&.B\&.A\&.in\-addr\&.arpa\&. See special note about IPSECKEY processing, below\&.
.SS "IPSECKEY6 queryid A:B::C:D"
.PP
This request looks up the IPSECKEY resource record found in the reverse map for the IPv6 address
\fBA:B::C:D\fR, i\&.e\&. it looks the 32\-nibble long entry in ip6\&.arpa (and ip6\&.int)\&. See special note about IPSECKEY processing, below\&.
.SS "OE4 queryid A\&.B\&.C\&.D"
.PP
This request looks an appropriate record for Opportunistic Encryption for the given IP address\&. This attempts to look for the delegation record\&. This may be one of IPSECKEY, KEY, or TXT record\&. Unless configured otherwise, (see OE4 Directives, below), then a query type of ANY will be used to retrieve all relevant records, and all will be returned\&.
.SS "OE6 queryid A:B::C:D"
.PP
This request looks an appropriate record for Opportunistic Encryption for the given IPv6 address\&. This attempts to look for the delegation record\&. This may be one of IPSECKEY, KEY, or TXT record\&. Unless configured otherwise, (see OE Directives, below), then a query type of ALL will be used to retrieve all relevant records, and all will be returned\&. i\&.e\&. it looks the 32\-nibble long entry in ip6\&.arpa (and ip6\&.int)\&.
.SS "A queryid FQDN"
.PP
This request looks up the A (IPv4) resource record for the given
\fBFQDN\&.\fR\&.
.SS "AAAA queryid FQDN"
.PP
This request looks up the AAAA (IPv6) resource record for the given
\fBFQDN\&.\fR\&.
.SS "VPN4"
.PP
synonomous for KEY4
.SS "VPN6"
.PP
synonomous for KEY6
.SH "REPLIES TO QUERIES"
.PP
All replies from the queries are in the following format:
.sp
.if n \{\
.RS 4
.\}
.nf

<ID> <TIME> <TTL> <TYPE> <TYPE\-SPECIFIC> \en

.fi
.if n \{\
.RE
.\}
.PP
\fIID\fR
.RS 4
this is the
\fBqueryid\fR
value that was provided in the query\&. It is repeated on every line to permit the replies to be properly associated with the query\&. When the response is not ascribable to particular query (such as for a mis\-formed query), then the query ID "0" will be used\&.
.RE
.PP
\fITIME\fR
.RS 4
this is the current time in seconds since epoch\&.
.RE
.PP
\fITTL\fR
.RS 4
for answers which have a time to live, this is the current value\&. The answer is valid for this number of seconds\&. If there is no useful value here, then the number 0 is used\&.
.RE
.PP
\fITYPE\fR
.RS 4
This is the type of the record that is being returned\&. The types are described in the next section\&. The TYPE specific data that follows is specific to the type\&.
.RE
.PP
The replies are limited to 4096 bytes, a value defined as
\fBLWDNSQ_RESULT_LEN_MAX\fR\&. This is defined in
\fIfreeswan\&.h\fR\&.
.PP
All of the replies which include resource records use the standard presentation format (with no line feeds or carriage returns) in their answer\&.
.SS "START"
.PP
This reply indicates that a query has been received and has been started\&. It serves as an anchor point for timing, as well as an acknowledgement\&.
.SS "DONE"
.PP
This reply indicates that a query is entirely over, and no further information from this query will be sent\&.
.SS "RETRY"
.PP
This reply indicates that a query is entirely over, but that no data was found\&. The records may exist, but appropriate servers could not be reached\&.
.SS "FATAL"
.PP
This reply indicates that a query is entirely over, and that no data of the type requested could be found\&. There were no timeouts, and all servers were available and confirmed non\-existances\&. There may be NXT records returned prior to this\&.
.SS "CNAME"
.PP
This is an interim reply, and indicates that a CNAME was found (and followed) while performing the query\&. The value of the CNAME is present in the type specific section\&.
.SS "CNAMEFROM"
.PP
This is an interim reply, and indicates that a CNAME was found\&. The original name that was queries for was not the canonical name, and this reply indicates the name that was actually followed\&.
.SS "NAME"
.PP
This is an interim reply\&. The original name that was queries for was not the canonical name\&. This reply indicates the canonical name\&.
.SS "DNSSEC"
.PP
This is an interim reply\&. It is followed either by "OKAY" or "not present\&. It indicates if DNSSEC was available on the reply\&.
.SS "TXT and AD\-TXT"
.PP
This is an interim reply\&. If there are TXT resource records in the reply, then each one is presented using this type\&. If preceded by AD\-, then this record was signed with DNSSEC\&.
.SS "A and AD\-A"
.PP
This is an interim reply\&. If there are A resource records in the reply, then each one is presented using this type\&. If preceded by AD\-, then this record was signed with DNSSEC\&.
.SS "AAAA and AD\-AAAA"
.PP
This is an interim reply\&. If there are AAAA resource records in the reply, then each one is presented using this type\&. If preceded by AD\-, then this record was signed with DNSSEC\&.
.SS "PTR and AD\-PTR"
.PP
This is an interim reply\&. If there are PTR resource records in the reply, then each one is presented using this type\&. If preceded by AD\-, then this record was signed with DNSSEC\&.
.SS "KEY and AD\-KEY"
.PP
This is an interim reply\&. If there are KEY resource records in the reply, then each one is presented using this type\&. If preceded by AD\-, then this record was signed with DNSSEC\&.
.SS "IPSECKEY and AD\-IPSECKEY"
.PP
This is an interim reply\&. If there are IPSEC resource records in the reply, then each one is presented using this type\&. If preceded by AD\-, then this record was signed with DNSSEC\&.
.SH "SPECIAL IPSECKEY PROCESSING"
.PP
At the time of this writing, the IPSECKEY resource record is not entirely specified\&. In particular no resource record number has been assigned\&. This program assumes that it is resource record number 45\&. If the file /etc/ipsec\&.d/lwdnsq\&.conf exists, and contains a line like
.sp
.if n \{\
.RS 4
.\}
.nf

ipseckey_rr=\fBnumber\fR

.fi
.if n \{\
.RE
.\}
.PP
then this number will be used instead\&. The file is read only once at startup\&.
.SH "OE DIRECTIVES"
.PP
If the file /etc/ipsec\&.d/lwdnsq\&.conf exists, and contains a line like
.sp
.if n \{\
.RS 4

CHANGES

Note that listings are in chronological order of release times, not in order
of version numbers, so you will find 2.5.x and 2.6.x releases intersperced.

See also docs/KNOWN_BUGS and http://bugs.openswan.org/

v2.6.30 (unreleased)
* Fix for: ASSERTION FAILED at connections.c:1579: c->kind == CK_TEMPLATE [Paul]
* Add --listen/listen= option to limit listening to a single IP [Paul]
* sa_policy_bit_names was missing ModeConfig DNS and WINS bit names [Paul]
* SAREF: Add sareftrack=<no|yes|conntrack> connection option [Paul]
* Fix for "handling event EVENT_RETRANSMIT for <invalid>" [Paul]
* Fix for specifying protport=47 (GRE has no ports) [Paul]
* Don't fill traffic selector struct in IKEv2 child SA when not recv'd [Paul]
* Add geode-aes to the list of crypto modules to load [Paul]
* Don't install ipsec init script as /etc/init.d/setup [David]
* Don't create rc.? symlinks - let user use chkconfig etc instead [Paul]
* Bugtracker bugs fixed:
   # 252/619 more than 20 payloads in message; ignored [paul]
   # 690 ipsec lwdnsq --help not implemented
   # 860 Port --random for newhostkey [Paul]
   #1005 Incorrect message "R_U_THERE_ACK has unexpected sequence number" [Mike]
   #1040: Fix to compile without DEBUG [Paul]
   #1054: Startup warning: "ignored obsolete keyword (null)" [Michael Smith]
   #1112: Prototypes only, if function enabled in c-source with KLIPS or
          PFKEY [Henry N.]
   #1115: Fix various warnings u_char * vs. char * for sscanf,... [Henry N.]
   #1149: pluto uses empty NAT_OA as IDci, Server behind NAT and non natted
          Windows XP [Wolfgang Nothdurft]
   #1151: The ipsec module is not removed by 'ipsec setup stop' [Paul]

v2.6.29 (September 27, 2010)
! This release is made for CVE-2010-3302 and CVE-2010-3308
! Duplicate CVE's issues for these are CVE-2010-3752 and CVE-2010-3753
* XAUTH: Avoid buffer overflow in CISCO DNS info [dhr/paul]
         Avoid shell problems with single quotes CISCO DNS paramters [dhr/paul]
  http://www.cve.mitre.org/cgi-bin/cvename.cgi?name=2010-3302
* XAUTH: Avoid buffer overflow in CISCO BANNER [dhr/paul]
         Avoid shell problems with single quotes in CISCO paramters [dhr/paul]
  http://www.cve.mitre.org/cgi-bin/cvename.cgi?name=2010-3308
* NETKEY: Fix for spurious %hold netlink-acquires [Paul/dhr]
* KLIPS: Fix compiling on 2.6.18 based RHEL5 kernels [Paul]
* Various fixes based on automated source code review [dhr]
* SAREF: Updated for 2.6.35 kernel [Harald]
* KLIPS: Updated for 2.6.35 kernel [Harald]
* PACKAGING Use Epoch 1: for Debian/Ubuntu [Simon]
* MAST: fix iptables rule "leak" on rekey [Bart]
* MAST: use only the most recent iptables rule [Bart]
* pluto: restrict rekeymargin to be smaller than salifetime [Bart]
* MAST: ensure we don't end up with mtu=0 on mast0 [Bart/Paul]
* MAST: enforce outgoing tunnel policy [Bart]
* MAST: use addflow pfkey command to set policy on tunnel SAs [Bart]
* Added a new pfkey flag, POLICYONLY, to the ADDFLOW command [Bart]
* MAST: allow for setting of policy for inbound SAs [Bart]
* MAST: favour deleting an SA even if the pfkey op failed [Bart]
* HAVESTATSD: Log new phase2 messages as a result of a rekey [Paul]
* MAST: use iptables --comment to show the conn name [Bart]
* VNET: differentiate instantiation of road warriors and vnet [Paul]
* Log LEAK_DETECTIVE and HAVE_LIBNSS support on startup [Paul]
* [IKEv2] connections were broken since 2.6.25 [Avesh]
* MAST: new "ipsec policy" command replaces "ipsec eroute" [Bart]
* Fix SElinux warning in realsetup (bz628879) [Avesh]
* Support for SHA2_256 in IKEv2 (bz621790) [Avesh]
* IKEv2: Fix for using MD5 and PRF conversion function [Avesh]
* SAREF: Improved workaround for rp_filter [Bart]
* NSS: Increase minimum nss for rhbz#453577 [Paul]
  (this allows us to revert workaround in git 6c8ff2791d1)
* SAREF: Added /proc/net/ipsec/saref that shows kernel patch state [Bart]
* SAREF: Add SAref checks to ipsec verify [bart/paul]
* A bunch of Makefile.inc variables could not be overridden properly [Paul]
* Bugtracker bugs fixed:
   #1042 NF_INET_LOCAL_OUT vs NF_IP_LOCAL_OUT still broken [David]
   #1102: Ensure KLIPS still works on older kernels [Paul]
   #1113: crypt_dh.c: Fix "warning: multi-line comment" [Tuomo]

v2.6.28 (July 29, 2010)
* MAST: Fix SAref vmalloc() call that could cause a kernel panic [Bart]
* SAREF: rework exported interface to avoid might_sleep() during rcu lock [Bart]
* SAREF: Use the correct mask when updating nfmark in rcv path [Bart]
* MAST: Cleanup updown.mast iptables rule management [Bart]
* MAST: Rework mast init scripts to use conntrack [Bart]
* MAST: Remove iptables rules after SA is deleted [Bart]
* SAREF: Fix bug in stream-socket saref mode [Bart]
* SAREF: Ported to apply on Linux 2.6.34 [Harald]
* MAST: Fix for NAT-T mode (set ixs->outgoing_said in mast mode) [Bart]

programs/lwdnsq/lwdnsq.c
/*
 * DNS KEY lookup helper
 * Copyright (C) 2002 Michael Richardson <mcr@freeswan.org>
 * 
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.  See <http://www.fsf.org/copyleft/gpl.txt>.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * for more details.
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h> 
#include <unistd.h> 

#include <errno.h>
#include <getopt.h>
#include <setjmp.h>
#include <ctype.h>
#include <signal.h>
#include <time.h>

#include <arpa/nameser.h>
#include <lwres/netdb.h>

#include <isc/lang.h>
#include <isc/magic.h>
#include <isc/types.h>
#include <isc/result.h>
#include <isc/mem.h>
#include <isc/buffer.h>
#include <isc/region.h>
#include <dns/name.h>
#include <dns/rdata.h>
#include <dns/rdatastruct.h>
#include <lwres/netdb.h>
#include <lwres/async.h>

#if defined(__CYGWIN__) || (defined(macintosh) || (defined(__MACH__) && defined(__APPLE__)))
#define getopt_long_only getopt_long
#endif

#include "lwdnsq.h"

FILE *cmdlog = NULL;

extern int EF_DISABLE_BANNER;
extern int EF_ALIGNMENT;
extern int EF_PROTECT_BELOW;
extern int EF_PROTECT_FREE;
extern int EF_ALLOW_MALLOC_0;
extern int EF_FREE_WIPES;


static void
usage(char *name)
{	
	fprintf(stdout,"usage: %s [-h] [-i] [-s] [-g] [-l <file>] [-X] [-Z]\n", name);
	fprintf(stdout," long opts: --help --prompt --serial --debug --log <file> --regress --ignoreeof\n");
	exit(1);
}

static struct option const longopts[] =
{
	{"help", 0, 0, 'h'},
	{"prompt", 0, 0, 'i'},
	{"serial", 0, 0, 's'},
	{"debug",  0, 0, 'g'},
	{"log",    1, 0, 'l'},
	{"regress",0, 0, 'X'},
	{"ignoreeof",0, 0, 'Z'},
	{0, 0, 0, 0}
};

/* globals */
jmp_buf getMeOut;

static void sig_handler(int sig)
{
  fprintf(stderr, "Caught signal %d, cleaning up and exiting\n", sig);
  longjmp(getMeOut, 1);
}

static void cmdprompt(dnskey_glob *gs)
{
	if(gs->prompt) {
		printf("lwdnsq> ");
	}
	fflush(gs->cmdproto_out);
}

static void quitprog(dnskey_glob *gs,
		     int argc UNUSED,
		     char **argv UNUSED)
{
	argc = argc;
	argv = argv;
	gs->done=1;
}

static void setdebug(dnskey_glob *gs,
	      int argc,
	      char **argv)
{
	if(argc > 1) {
		gs->debug=strtoul(argv[1],NULL,0);
	}
	printf("0 DEBUG is %d\n",gs->debug);
}


static int cmdparse(dnskey_glob *gs,
		    char *cmdline)
{
	char *argv[256];
	int   argc;
	char *arg;
	static const struct cmd_entry {
		const char *cmdname;
		void (*cmdfunc)(dnskey_glob *, int, char **);
	} cmds[]={
		{"key",       lookup_key},
		{"key4",      lookup_key4},
		{"key6",      lookup_key6},
		{"txt",       lookup_txt},
		{"txt4",      lookup_txt4},
		{"txt6",      lookup_txt6},
		{"ipseckey",  lookup_ipseckey},
		{"ipseckey4", lookup_ipseckey4},
		{"ipseckey6", lookup_ipseckey6},
		{"oe4",       lookup_oe4},
		{"oe6",       lookup_oe6},
		{"vpn4",      lookup_key4},
		{"vpn6",      lookup_key6},
		{"quit",      quitprog},
		{"a",         lookup_a},
		{"aaaa",      lookup_aaaa},
		{"debug",     setdebug},
		{NULL,        NULL}};
	const struct cmd_entry *ce = cmds;

	if(cmdlog != NULL) {
		fprintf(cmdlog, "%lu|%s\n", (unsigned long)time(NULL),cmdline);
		fflush(cmdlog);
	}

	argc=0;
	
	/* skip initial spaces */
	while(cmdline && isspace(*cmdline)) {
		cmdline++;
	}

	while(cmdline && *cmdline!='\0' &&
	      (arg=strsep(&cmdline, " \t\n"))!=NULL) {
	  if (argc < (signed)(sizeof(argv)/sizeof(*argv - 1))) {
	    /* ignore arguments that would overflow.
	     * XXX should generate a diagnostic.
	     */
	    argv[argc++]=arg;
	  }
	  while(cmdline && isspace(*cmdline)) {
	    cmdline++;
	  }
	}
	argv[argc]=NULL;

	if(argc==0 || argv[0][0]=='\0') {
	    /* ignore empty line */
	} else if(strcasecmp("help", argv[0]) == 0) {
	    fprintf(gs->cmdproto_out, "0 HELP\n");
	    for (; ce->cmdname != NULL; ce++)
		fprintf(gs->cmdproto_out, "0 HELP %s\n", ce->cmdname);
	} else {
	    for (;; ce++) {
		if (ce->cmdname == NULL) {
		    fprintf(gs->cmdproto_out, "0 FATAL unknown command \"%s\"\n", argv[0]);
		    break;
		}
		if(strcasecmp(ce->cmdname, argv[0])==0) {
		    (*ce->cmdfunc)(gs, argc, argv);
		    break;
		}
	    }
	}

	if (!gs->done) {
		gs->promptnow = 1;
	}
	return 0;
}

static int cmdread(dnskey_glob *gs,
		   char  *buf,
		   int    len)
{
	char *nl;
	int   cmdlen;

	cmdlen=0;

	/* 
	 * have to handle partial reads and multiple commands
	 * per read, since this may in fact be a file or a pipe.
	 */
	if((gs->cmdloc + len + 1) > sizeof(gs->cmdbuf)) {
		fprintf(stderr, "command '%.*s...' is too long, discarding!\n",
			40, buf);
		fflush(stdout);
		
		gs->cmdloc=0;
		return 0;
	}
	memcpy((unsigned char *)gs->cmdbuf+gs->cmdloc, (unsigned char *)buf, len);
	gs->cmdloc+=len;
	gs->cmdbuf[gs->cmdloc]='\0';

	while((nl = strchr((char *)gs->cmdbuf, '\n')) != NULL) {
		/* found a newline, so turn it into a \0, and process the
		 * command, and then we will pull the rest of the buffer
		 * up.
		 */
		*nl='\0';
		cmdlen= nl - gs->cmdbuf +1;

		cmdparse(gs, (char *)gs->cmdbuf);

		gs->cmdloc -= cmdlen;
		memmove((unsigned char *)gs->cmdbuf, (unsigned char *)gs->cmdbuf+cmdlen, gs->cmdloc);
	}
	return 1;
}

int
main(int argc, char *argv[])
{
	char *program_name;
	dnskey_glob gs;
	int c;
	static int ignoreeof=0;  /* static to avoid longjmp clobber */
	int ineof;
	char *logfilename=NULL;

	memset(&gs, 0, sizeof(dnskey_glob));

#if 0
	printf("PID: %d\n", getpid());
	sleep(60);
#endif

#ifdef EFENCE
	EF_DISABLE_BANNER=1;
	/* EF_ALIGNMENT=4; */
	EF_PROTECT_BELOW=1;
	EF_PROTECT_FREE=1;
	/* EF_ALLOW_MALLOC_0; */
	EF_FREE_WIPES=1;
#endif

	{ 
			FILE *newerr;
			newerr = fopen("/var/run/pluto/lwdnsq.log", "a+");
			if(newerr) {
				close(2);
				dup2(fileno(newerr), 2);
				fclose(newerr);
			}
			fprintf(stderr, "stderr capture started\n");
			setbuf(stderr, NULL);
	}


	program_name = argv[0];
	gs.concurrent = 1;

	if(lwres_async_init(&gs.lwctx) != ERRSET_SUCCESS) {
		fprintf(stderr, "Can not initialize async context\n");
		exit(3);
	}

	if(isc_mem_create(0,0,&gs.iscmem) != ISC_R_SUCCESS) {
		fprintf(stderr, "Can not initialize isc memory allocator\n");
		exit(4);
	}

	if(isc_buffer_allocate(gs.iscmem, &gs.iscbuf, LWDNSQ_RESULT_LEN_MAX)) {
		fprintf(stderr, "Can not allocate a result buffer\n");
		exit(5);
	}

	while((c = getopt_long_only(argc, argv, "hdgl:siXZ", longopts, 0)) != EOF) {
		switch(c) {
		case 'h':
			usage(program_name);
			break;
		case 'd':
			gs.debug+=2;
			logfilename="/var/run/pluto/lwdns.req.log";
			break;

		case 'g':
			gs.debug++;
			break;

		case 's':
			gs.concurrent=0;
			break;
		case 'i':
			gs.prompt=1;
			break;
		case 'X':
			gs.regress++;
			break;

		case 'l':
			logfilename=optarg;
			break;

		case 'Z':
			ignoreeof=1;
			break;

		default:
			usage(program_name);
			break;
		}
	}

	if(gs.debug && ignoreeof) {
		fprintf(stderr, "Ignoring end of file\n");
	}

	if(isatty(0)) {
		gs.prompt=1;
	}

	if(logfilename != NULL) {
	  cmdlog = fopen(logfilename, "w");
	  if(cmdlog == NULL) {
	    fprintf(stderr, "Can not open %s: %s\n", logfilename, strerror(errno));
	  }
	}

	/* do various bits of setup */
	if(setjmp(getMeOut)!=0) {
		signal(SIGINT,  SIG_DFL);
		signal(SIGPIPE, SIG_IGN);
		
		/* cleanup_crap(); */
		
		exit(1);
	}
	
	if(signal(SIGINT, sig_handler) == SIG_ERR)
		perror("Setting handler for SIGINT");
	
	if(signal(SIGPIPE, sig_handler) == SIG_ERR)
		perror("Setting handler for SIGINT");
	
	cmdprompt(&gs);

	ineof = 0;
	gs.done = 0;
	gs.cmdproto_out = stdout;

	while(!gs.done) 
	{
		int    timeout;
		char   buf[128];
		int    n;
		int    rlen, err2;

		timeout = LWDNSQ_RETRANSMIT_INTERVAL * 1000;

		/* 
		 * do not prompt for more items, or look for HUP if
		 * we are in serial mode.. We know we are working
		 * if there are questions in flight. 
		 */
		gs.l_fds[1].events = 0;

		if(gs.concurrent || gs.dns_inflight == 0) {
			if(gs.promptnow && gs.prompt) {
				cmdprompt(&gs);
				gs.promptnow = FALSE;
			}
			gs.l_fds[1].events = POLLIN;
			if(!ignoreeof) {
				gs.l_fds[1].events |= POLLHUP;
			}
			gs.l_nfds = 2;
		} else {
			gs.l_fds[1].events = 0;
			gs.l_nfds = 1;
		}
			
		gs.l_fds[1].revents = 0;

		gs.l_fds[0].events = POLLIN|POLLHUP;
		gs.l_fds[0].revents = 0;
		gs.l_fds[0].fd = lwres_async_fd(gs.lwctx);

		if(gs.debug > 1) {
			fprintf(stderr, "=== invoking poll(,%d,%d)\n",
				gs.l_nfds, timeout);
			for(n = 0; n < gs.l_nfds; n++) {
				fprintf(stderr, "=== waiting on fd#%d\n",
					gs.l_fds[n].fd);
			}
			fprintf(stderr, "=== inflight: %d\n", gs.dns_inflight);
		}

		errno = 0;
		n = poll(gs.l_fds, gs.l_nfds, timeout);
		err2 = errno;

		/* there was some activity, so look for pending retransmits
		 * first, even if it wasn't due to the timeout.
		 */
		{
			struct dnskey_lookup *dl, *dlthis;
			
			dl = gs.dns_outstanding;
			
			while(dl) {
				/* dl could go away */
				dlthis = dl;
				dl = dl->next;

				dns_rexmit(&gs, dlthis, 0);
			}
		}

		if(n < 0) {
			perror("poll");
		}

		if(gs.debug > 1) {
			fprintf(stderr, "=== poll returned with %d\n", n);
		}
				
		while(n>0) {
			if((gs.l_fds[0].revents & POLLERR) == POLLERR ||
			   (gs.l_fds[1].revents & POLLERR) == POLLERR)
			{
				struct dnskey_lookup *dl, *dlthis;
				if(gs.debug > 1) {
					fprintf(stderr, "=== poll returned error %s\n",
						strerror(err2));
				}

				/* on error, we force retransmit, because
				 * one reasonable source of errors it because
				 * the named isn't ready yet.
				 */
				sleep(1);
			
				dl = gs.dns_outstanding;
				while(dl) {
					/* dl could go away */
					dlthis = dl;
					dl = dl->next;
					
					dns_rexmit(&gs, dlthis, 1);
				}
				n = 0;
				break;
			}

			/* see if there are DNS events coming back */
			if((gs.l_fds[0].revents & POLLIN) == POLLIN) {
				if(gs.debug > 1) {
					fprintf(stderr,
						"=== new responses from lwdnsd\n");
				}

				process_dns_reply(&gs);
				fflush(stdout);
				n--;
			}

			if(!ignoreeof &&
			   (gs.l_fds[1].revents & POLLHUP) == POLLHUP)
			{
				break;
			}

			if((gs.l_fds[1].revents & POLLIN) == POLLIN) {
				
				rlen=read(0, buf, sizeof(buf));

				if(gs.debug > 1) {
					if(rlen > 0) {
						buf[rlen]='\0';
					}
					fprintf(stderr,
						"=== new commands on fd 0: %d: %s\n",
						rlen, buf);
				}

				if(rlen > 0) {
					cmdread(&gs, buf, rlen);
				} else if(rlen == 0) {
					ineof = 1;
					if(!ignoreeof) {
						/* EOF, die */
						gs.done=1;
					}
				}
				n--;
			} 

		}

		if((gs.l_fds[1].revents & POLLHUP) == POLLHUP)
		{
			ineof = 1;
			if(!ignoreeof)
			{
				gs.done=1;
			}
		}

		if(ignoreeof) {
			/* if we have exhausted the input,
			 * and there are none in flight,
			 * then exit, finally.
			 */
			if(ineof) { 
				if(gs.dns_inflight == 0) {
					gs.done=1;
				}
			}
		}

		if(gs.debug) {
			fprintf(stderr, "=== ineof: %d inflight: %d\n",
				ineof, gs.dns_inflight);
		}

	}

	signal(SIGINT,  SIG_DFL);
	signal(SIGPIPE, SIG_IGN);
  
	exit(0);
}
	
/*
 * $Log: lwdnsq.c,v $
 * Revision 1.23  2005/08/26 19:13:48  mcr
 * 	fixed attempt to write an #ifdef.
 *
 * Revision 1.22  2005/08/25 01:24:40  paul
 * Added darwin target for MacOSX compile. Some additional or changed ifdef's
 * to make it compile
 *
 * Revision 1.21  2005/08/05 17:23:29  mcr
 * 	adjustment of signed/unsigned issues for gcc4-cygwin.
 *
 * Revision 1.20  2005/08/05 01:42:21  mcr
 * 	getopt_long_only is GNU-ism.
 *
 * Revision 1.19  2004/12/02 06:16:19  mcr
 * 	fixed long standing bug with async resolver when there was
 * 	more than one outstanding request.
 *
 * Revision 1.18  2004/11/25 18:16:47  mcr
 * 	make sure to actually open the log file.
 *
 * Revision 1.17  2004/11/24 01:36:10  mcr
 * 	if debugging is on, then log all requests that are sent,
 * 	so that we can later replay, looking for bugs.
 *
 * Revision 1.16  2003/12/02 04:34:09  mcr
 * 	if lwresd does not start early enough, then the request is
 * 	lost and never retransmitted. This version can now deal with
 * 	named being slow to start, or taking along time to verify
 * 	DNSSEC signatures.
 * 	This also fixes the non-concurrent stuff.
 *
 * Revision 1.15  2003/12/02 02:55:19  mcr
 * 	prompting in non-concurrent mode looks right now.
 * 	do not ask for EOF messages from stdin if we are in ignoreeof
 * 	mode.
 *
 * Revision 1.14  2003/12/01 21:45:17  mcr
 * 	the lwctx structure needs to be properly initialized.
 * 	the lwdnsq program needs to have the --serial option finished,
 * 	and a retransmitter installed.
 *
 * Revision 1.13  2003/09/19 02:56:11  mcr
 * 	fixes to make lwdnsq compile with strictest compile flags.
 *
 * Revision 1.12  2003/09/16 05:01:14  mcr
 * 	prefix all debugging with === so that it can be easily removed.
 *
 * Revision 1.11  2003/09/10 04:43:52  mcr
 * 	final fixes to lwdnsq to exit only when all requests are done,
 * 	and we have been told to wait, *OR* if there is an EOF in stdin.
 *
 * Revision 1.10  2003/09/03 01:13:24  mcr
 * 	first attempt at async capable lwdnsq.
 *
 * Revision 1.9  2003/04/02 07:37:57  dhr
 *
 * lwdnsq: fix non-deterministic bug in handling batched input
 *
 * Revision 1.8  2003/02/08 04:03:06  mcr
 * 	renamed --single to --serial.
 *
 * Revision 1.7  2003/01/14 03:01:14  dhr
 *
 * improve diagnostics; tidy
 *
 * Revision 1.6  2002/12/19 07:29:47  dhr
 *
 * - avoid (improbable) buffer overflow
 * - suppress prompt after "quit" command
 * - add space to prompt to match aesthetics and man page
 * - elminate a magic number
 *
 * Revision 1.5  2002/12/19 07:08:42  dhr
 *
 * continue renaming dnskey => lwdnsq
 *
 * Revision 1.4  2002/12/12 06:03:41  mcr
 * 	added --regress option to force times to be regular
 *
 * Revision 1.3  2002/11/25 18:37:48  mcr
 * 	make sure that we exit cleanly upon EOF.
 *
 * Revision 1.2  2002/11/16 02:53:53  mcr
 * 	lwdnsq - with new contract added.
 *
 * Revision 1.1  2002/10/30 02:25:31  mcr
 * 	renamed version of files from dnskey/
 *
 * Revision 1.3  2002/10/09 20:14:16  mcr
 * 	make sure to flush stdout at the right time - do it regardless
 * 	of whether or not we are printing prompts.
 *
 * Revision 1.2  2002/09/30 18:55:54  mcr
 * 	skeleton for dnskey helper program.
 *
 * Revision 1.1  2002/09/30 16:50:23  mcr
 * 	documentation for "dnskey" helper
 *
 * Local variables:
 * c-file-style: "linux"
 * c-basic-offset: 2
 * End:
 *
 */
